controllers/edgedeployment_controller.go

```diff
@@ -64,71 +66,160 @@ func (r *EdgeDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		}
 		return ctrl.Result{Requeue: true}, err
 	}
-	edgeDevice, err := r.EdgeDeviceRepository.Read(ctx, edgeDeployment.Spec.Device, edgeDeployment.Namespace)
-	if err != nil {
-		return ctrl.Result{Requeue: true}, err
+
+	if edgeDeployment.DeletionTimestamp == nil && !utils.HasFinalizer(&edgeDeployment.ObjectMeta, YggdrasilDeviceReferenceFinalizer) {
+		deploymentCopy := edgeDeployment.DeepCopy()
+		deploymentCopy.Finalizers = []string{YggdrasilDeviceReferenceFinalizer}
+		err := r.EdgeDeploymentRepository.Patch(ctx, edgeDeployment, deploymentCopy)
+		if err != nil {
+			return ctrl.Result{Requeue: true}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	err = r.addOwnerReference(ctx, edgeDevice, edgeDeployment)
+	labelledDevices, err := r.getLabelledEdgeDevices(ctx, edgeDeployment.Name, edgeDeployment.Namespace)
 	if err != nil {
-		return ctrl.Result{Requeue: true}, err
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Cannot retrieve Labelled Edge Deployments")
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	if edgeDeployment.DeletionTimestamp != nil {
+		if utils.HasFinalizer(&edgeDeployment.ObjectMeta, YggdrasilDeviceReferenceFinalizer) {
+			err = r.finalizeRemoval(ctx, labelledDevices, edgeDeployment)
+			if err != nil {
+				return ctrl.Result{Requeue: true}, err
+			}
+		}
+		return ctrl.Result{}, nil
 	}
 
-	if edgeDeployment.Status.Phase == "" {
-		edgeDeployment.Status.Phase = managementv1alpha1.Deploying
-		edgeDeployment.Status.LastTransitionTime = metav1.Now()
+	edgeDevices, err := r.getMatchingEdgeDevices(ctx, edgeDeployment)
```

Comment:
I think it would be good to skip if no changes to `Device` or `DeviceSelector` detected.

Reply:
I don't think we can detect that without getting previous version of the CR. Besides, fetching the matching devices allows us to add/remove deployments to/from _devices_ that changed.

Reply:
I am not sure whether it would work in this case but there is a way to [filter](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) specific events based on new vs old CR.

Reply:
I think filters should work in this case, but there is still the second argument for keeping it - a little better deployment consistency.